spec/unit/util/crypto_spec.rb

```diff
@@ -74,20 +74,26 @@
 
   context 'encrypts & decrypt' do
     let(:string) { random_str }
+    let(:empty_string) { '' }
 
     specify '#encrypts and decrypts a string' do
+      expect(string).to be_ascii_only
       encrypted = subject.encrypt(string)
-      expect(subject.decrypt(encrypted)).to eql(string)
+      expect(encrypted).to be_truthy
+      decrypted = subject.decrypt(encrypted)
+      expect(decrypted).to eql(string)
+      expect(decrypted).to be_ascii_only
     end
 
-  end
-
-  context 'encrypting an empty string' do
```

Comment:
While I agree I commits for this PR are not in required order. 
https://github.com/ably/ably-ruby/commit/a0a3b98c997944abeb8edb7adc5157bf0188c1eb was aimed at fixing old ruby versions. 
That said, I don't have old version of ruby installed on my local machine. So, there wasn't a way for me to test the behaviour if it will work or not unless I push it on the CI.
I am not a big fan of force push. Also, being new to ruby there are tests being updated in different iterations. I wish, I could do all of it in one meaningful commit : |
I would not recommend to go through each commit. Rather the final PR itself which represents final changes.

PS. Even I don't remember commit related to each line of code. So, it will be difficult for both of us to discuss why changes were made at the time.

I also agree that we need to follow [best commit guidelines](https://github.com/ably/ably-ruby/pull/402#issuecomment-2135850252). I will try doing the same in subsequent commits. That said, I would like you to review code based on the final PR, since it's relatively small.

Reply:
thanks for the review !

Reply:
> That said, I don't have old version of ruby installed on my local machine. So, there wasn't a way for me to test the behaviour if it will work or not unless I push it on the CI.

That's fine, nothing wrong with using CI for experimenting with old Ruby versions 👍 

> I am not a big fan of force push

The fact is that there's pretty much a 0% chance of getting the commits right on your first try. As you say, software development is an iterative process. Whilst working on a feature we try things, we make mistakes, we fix things. This is impossible to avoid. It's hence extremely difficult to naturally get a good commit history; a good commit history is something that you have to explicitly craft. This means needing to embrace the tools that allow you to rewrite commit history.

> Even I don't remember commit related to each line of code. So, it will be difficult for both of us to discuss why changes were made at the time.

I don't understand this. As a reviewer of a PR, I think it's perfectly reasonable to want to understand the motivation for every change that the PR introduces.

>  I would not recommend to go through each commit. Rather the final PR itself which represents final changes.

But then the PR needs to come with good explanations of all of the changes that have been made and the motivation for these changes, which this PR didn't.

Reply:
Understood! From now onwards, I will try to make sure meaningful commits are added. Along with this, I will add relevant description to the PR 👍 
Thanks for the review again !